### PR TITLE
feat(agent-runtime): handle JSON in read_url with shape, preview, and drill-in

### PR DESCRIPTION
**Pull request by kristofferremback**
## Problem

The `read_url` agent tool (used by Ariadne and other personas) only handled HTML and plain text. For JSON it did two unhelpful things:

- **Rejected it outright** — `application/json` returned `"Unsupported content type. Only HTML and plain text are supported."`, so the agent couldn't read JSON APIs at all.
- **Truncated mid-structure** — the 50k-char fallback sliced the serialized string, which for JSON produces invalid, un-navigable text. The model can neither parse it nor tell what was cut.

We want the agent to read JSON: small payloads loaded in full, and large payloads inspected partially — see the shape, sample the values, and drill into the part that matters.

## Solution

`read_url` now treats JSON as a first-class content type.

- **Small JSON** → returned verbatim as `{ kind: "json", content: <object> }`.
- **Large JSON** (compact serialization over `MAX_CONTENT_LENGTH` = 50k chars) → instead of a truncated blob, returns three views:
  - `shape` — a recursive schema skeleton. Objects stay nested for scanning; arrays/primitives collapse to one-line descriptors like `"array[1043] of {id:string, name:string, price:number}"`.
  - `preview` — a *structural* sample that stays valid JSON: first few array items + `"…(N more items)"`, capped object keys, strings truncated with the original length noted.
  - `hint` + `byteSize` — tells the model the scale and exactly how to drill in.
- **Drill-in** → an optional `select` path argument. The model calls `read_url` again with e.g. `select=".data.items[0:20]"`; the tool re-fetches, navigates to that subtree, and re-applies the size logic.

### How it works

```
fetch → content-type?
  ├─ text/html  → markdown (unchanged)
  ├─ text/plain → text, OR JSON if the body actually parses (unchanged path otherwise)
  └─ application/json (+ *+json, or any body when `select` is given)
        parse → apply `select` if present →
          compact size ≤ 50k ? return full `content`
                              : return { shape, preview, hint, byteSize, truncated:true }
```

The drill-in is **stateless** — `select` re-fetches the URL rather than caching the body. That matches how the tool already works (every call re-fetches) and is correct for idempotent GET APIs; a per-turn body cache is a clean future optimization if re-fetch cost ever matters.

### Key design decisions

**1. Shape + structural preview instead of string truncation**

Slicing serialized JSON yields corrupt, un-parseable output. A schema skeleton plus a sampled-but-valid preview lets the model understand the whole structure and request exactly the branch it wants. This is the genuinely useful (and hard) part of the feature.

**2. Minimal custom path selector, not full jq or a JSONPath dependency**

`select` supports `.key`, `["quoted key"]`, `[index]` (negatives allowed), `[start:end]` slices, and `[*]` wildcards — pure JS, zero new dependencies, fully controlled and safe. Real jq would require a native binary in the runtime image; JSONPath would add a dependency for power we don't need yet.

**3. Direct paths are strict; wildcards are tolerant**

A missing key on a direct path returns an actionable error naming the keys that *do* exist (great self-correction signal for the model). A `[*]` wildcard null-fills elements missing the sub-path (heterogeneous arrays with optional fields are common — jq-like), but still errors if *every* element fails, so a wholesale-wrong path doesn't silently return all-null.

**4. JSON-aware `Accept` header**

Added `application/json;q=0.9` so content-negotiating APIs prefer serving JSON, while HTML stays top priority for normal web pages.

## New files

| File | Purpose |
| --- | --- |
| `packages/agent-runtime/src/tools/json-inspect.ts` | Pure helpers: `describeShape`, `structuralPreview`, `applySelect` (path parser + resolver) |
| `packages/agent-runtime/src/tools/json-inspect.test.ts` | 23 unit tests for the pure helpers |

## Modified files

| File | Change |
| --- | --- |
| `packages/agent-runtime/src/tools/read-url-tool.ts` | Accept JSON content types (+`*+json`, JSON-as-text/plain); add optional `select` arg; branch small-vs-large on compact serialized size; JSON-aware `Accept` header |
| `packages/agent-runtime/src/tools/read-url-tool.test.ts` | Added JSON handling tests (full, large summary, drill-in, bad-select error, content-type edge cases) |
| `apps/backend/src/features/agents/companion/prompt/system-prompt.ts` | Document JSON reading + `select` usage for the companion persona |

## Test plan

- [x] `bun test` in `packages/agent-runtime` — 127 tests pass (58 in the two touched files, incl. 25 new)
- [x] `bun run typecheck` — clean (agent-runtime + backend)
- [x] Pre-PR multi-perspective review (spec / correctness / design); all surviving findings addressed: wildcard tolerance, compact-size gate, `Accept` header, test assertion
- [ ] Manual: point Ariadne at a large JSON API, confirm it gets `shape`/`preview`/`hint` then drills in with `select`

---

🤖 _PR by [Claude Code](https://claude.com/claude-code)_

---
<a href="https://app.blacksmith.sh/threahq/codesmith/threa/pr/697"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with Codesmith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1782681361&installation_id=129946197&pr_number=697&repository=threahq%2Fthrea&return_to=https%3A%2F%2Fgithub.com%2Fthreahq%2Fthrea%2Fpull%2F697&signature=9fb4055f9b2a9e607d4c6e113c3e295a859de5b290592d100135732cbd4ba9a2"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with Codesmith</picture></a>
<sup>Need help on this PR? Tag <code>@codesmith</code> with what you need. Autofix is disabled.</sup>

### CODE REVIEW

File: packages/agent-runtime/src/tools/read-url-tool.ts [hunk 1]
```diff
@@ -232,24 +241,40 @@ export function createReadUrlTool() {
         }
 
         const contentType = response.headers.get("content-type") || ""
-        if (!contentType.includes("text/html") && !contentType.includes("text/plain")) {
+        const isHtml = contentType.includes("text/html")
+        const isPlain = contentType.includes("text/plain")
+        const declaresJson = /\bjson\b/i.test(contentType) || contentType.includes("+json")
+
```
[the remark this answers is not in the document]
Comment (kristofferremback) [reply packages/agent-runtime/src/tools/read-url-tool.ts]: Good catch — fixed in 9084a896. Media types are case-insensitive per RFC 9110, so the content-type is now lowercased once before the `isHtml`/`isPlain`/`declaresJson` checks. Added a test covering `Application/JSON; charset=UTF-8`.

File: packages/agent-runtime/src/tools/json-inspect.ts [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+    if (depth >= INLINE_MAX_DEPTH) return `array[${value.length}]`
+    return `array[${value.length}] of ${inlineShape(value[0], depth + 1)}`
+  }
+  if (isPlainObject(value)) {
+    if (depth >= INLINE_MAX_DEPTH) return "object"
+    const keys = Object.keys(value)
+    const shown = keys.slice(0, INLINE_MAX_KEYS)
+    const parts = shown.map((k) => `${k}:${inlineShape(value[k], depth + 1)}`)
+    if (keys.length > shown.length) parts.push("…")
+    return `{${parts.join(", ")}}`
```
[the remark this answers is not in the document]
Comment (kristofferremback) [reply packages/agent-runtime/src/tools/json-inspect.ts:59]: Agreed — fixed in 9084a896. Added a `summarizeKeyName` helper (cap `SUMMARY_KEY_MAX = 120`) applied in `inlineShape`, `describeShape`, and `structuralPreview`, so an extremely long property name is truncated with a `…(N chars)` marker instead of being copied verbatim. The large-payload guard now bounds both key count and key length. Added a test for a 500-char key.
